shell/platform/darwin/ios/framework/Headers/FlutterEngine.h

```diff
@@ -133,6 +133,34 @@ FLUTTER_EXPORT
  */
 - (void)destroyContext;
 
+/**
+ * Ensures that Flutter will generate a semantics tree.
+ *
+ * This is enabled by default if certain accessibility services are turned on by
+ * the user, or when using a Simulator. This method allows a user to turn
+ * semantics on when they would not ordinarily be generated and the performance
+ * overhead is not a concern, e.g. for UI testing. Note that semantics should
+ * never be programatically turned off, as it would potentially disable
+ * accessibility services an end user has requested.
+ *
+ * This method must only be called after launching the engine via
+ * `-runWithEntrypoint:` or `-runWithEntryPoint:libraryURI`.
+ */
+- (void)ensureSemanticsEnabled;
+
+/**
+ * Registers a callback that will be called once when semantics are available.
+ *
+ * If semantics are disabled, this callback will be stored but will not fire
+ * until they are enabled and the tree has been updated. Consider calling
+ * `-ensureSemanticsEnabled` first.
```

Comment:
One potential problem is that if a caller does call `-ensureSemanticsEnabled` first, what if the semantics tree is built before the callback is registered?

Calling `-ensureSemanticsEnabled` *after* registering the callback seems safer, but there would still be a problem if semantics are already enabled.

Reply:
Agreed - however, in my testing once semantics are enabled this fires pretty regularly.

Reply:
Do you know what triggers that?  Is it because the UI hasn't settled down, or we're rebuilding the semantics tree periodically even when it's not necessary?

Reply:
I'm concerned about this too but not entirely sure how to resolve it.  If I ask the Accessibility Bridge if it has semantics, it could tell me no, then we get interrupted by the UI thread with an update, and I go on to register the callback which thinks there was no data when by now there is. I really dont' want to put a mutex in there for this because this is performance sensitive stuff that happens a lot.  We could try to do a spin lock around it though.

If I try to force the framework to do a semantics update (as long as semantics are enabled), it's also possible that will someday get optimized out by a well intentioned person.

Reply:
I believe it's a combination of both.

Reply:
It seems weird to me if the semantics tree is periodically updating when nothing has changed.

I like the approach of forcing an update.  If you made, say, a forceSemanticsUpdate function and documented what it's used for, then I wouldn't be too worried about somebody unwittingly removing it.

Reply:
Would caching the signal help with this?  @GaryQian just addressed an independent but similar kind of problem.  The very first lifecycle message from Android to Flutter might be missed due to timing.  In that case, Gary used the Window object to cache the state to avoid losing it.  I'm not suggesting the Window for this case, necessarily, but maybe a cache within the C++ would eliminate race conditions here?